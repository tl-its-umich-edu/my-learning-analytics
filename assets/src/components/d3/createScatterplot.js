import * as d3 from 'd3'
import { adjustViewport } from '../../util/chart'
import { margin } from '../../constants/chartConstants'

function createScatterplot ({ data, width, height, el, tip, xAxisLabel, yAxisLabel }) {
  const [aWidth, aHeight] = adjustViewport(width, height, margin)

  const x = d3.scaleLinear()
    .domain([0, 100])
    .range([margin.left, aWidth - margin.right])

  const y = d3.scaleLinear()
    .domain([0, 100])
    .range([aHeight - margin.bottom, margin.top])

  const svg = d3.select(el).append('svg')
    .attr('width', aWidth)
    .attr('height', aHeight)

  const circle = svg.selectAll('.dot')
    .data(data).enter()
    .append('circle')
    .attr('cx', d => x(d.x))
    .attr('cy', d => y(d.y))
    .attr('r', d => 3)

  const xAxis = g => g
    .attr(`transform`, `translate(0, ${aHeight - margin.bottom})`)
    .call(d3
      .axisBottom(x)
      .tickSizeOuter(0)
      .tickValues([0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100])
      .tickFormat(d => `${d}%`)
    )
    .call(g => g.append('text')
      .attr('x', aWidth - margin.right)
      .attr('y', -4)
      .attr('fill', '#000')
      .attr('text-anchor', 'end')
      .text(xAxisLabel)
    )

  const yAxis = g => g
    .attr('transform', `translate(${margin.left},0)`)
    .attr('class', 'axis')
    .call(d3.axisLeft(y).tickSizeInner(-aWidth).ticks(6))
    .call(g => g.select('.domain').remove())
    .call(g => g.select('.tick:last-of-type text').clone()
      .attr('x', 5)
      .attr('fill', '#000')
      .attr('text-anchor', 'start')
      .text(yAxisLabel).attr('dy', -4)
    )

  svg.append('g')
    .call(xAxis)

  svg.append('g')
    .call(yAxis)

  if (tip) {
    svg.call(tip)
    circle
      .on('mouseover', tip.show)
      .on('mouseout', tip.hide)
  }
}

export default createScatterplot
