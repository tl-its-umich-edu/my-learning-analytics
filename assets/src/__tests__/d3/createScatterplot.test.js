/* global describe, test, expect */
import createScatterplot from '../../components/d3/createScatterplot'

const scatterplotData = Object.freeze(
  [
    {
      'x': 10,
      'y': 10
    },
    {
      'x': 10,
      'y': 11
    },
    {
      'x': 11,
      'y': 10
    },
    {
      'x': 30,
      'y': 30
    },
    {
      'x': 70,
      'y': 66
    },
    {
      'x': 90,
      'y': 90
    },
    {
      'x': 100,
      'y': 89
    },
    {
      'x': 40,
      'y': 77
    },
    {
      'x': 60,
      'y': 69
    }
  ]
)

describe('createScatterplot', () => {
  test('should build a scatter plot', () => {
    const div = document.createElement('div')
    createScatterplot({ data: scatterplotData, width: 1000, height: 500, el: div })
    expect(div).toMatchSnapshot()
  })
})
