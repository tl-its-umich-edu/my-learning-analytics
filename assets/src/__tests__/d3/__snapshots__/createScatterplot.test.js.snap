// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScatterplot should build a scatter plot 1`] = `
<div>
  <svg
    height="450"
    width="940"
  >
    <circle
      cx="128"
      cy="380"
      r="3"
    />
    <circle
      cx="128"
      cy="376"
      r="3"
    />
    <circle
      cx="136.8"
      cy="380"
      r="3"
    />
    <circle
      cx="304"
      cy="300"
      r="3"
    />
    <circle
      cx="656"
      cy="156"
      r="3"
    />
    <circle
      cx="832"
      cy="60"
      r="3"
    />
    <circle
      cx="920"
      cy="64"
      r="3"
    />
    <circle
      cx="392"
      cy="112"
      r="3"
    />
    <circle
      cx="568"
      cy="144"
      r="3"
    />
    <g
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 420)"
    >
      <path
        class="domain"
        d="M40.5,0.5H920.5"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(40.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(128.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          10%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(216.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(304.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          30%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(392.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          40%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(480.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          50%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(568.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          60%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(656.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          70%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(744.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          80%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(832.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          90%
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(920.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          100%
        </text>
      </g>
      <text
        fill="#000"
        text-anchor="end"
        x="920"
        y="-4"
      />
    </g>
    <g
      class="axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <g
        class="tick"
        opacity="1"
        transform="translate(0,420.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,340.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,260.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          40
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,180.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          60
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,100.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          80
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,20.5)"
      >
        <line
          stroke="currentColor"
          x2="940"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-3"
        >
          100
        </text>
        <text
          dy="-4"
          fill="#000"
          text-anchor="start"
          x="5"
        />
      </g>
    </g>
  </svg>
</div>
`;
